### PR TITLE
Add sequence packing support for hybrid model

### PR DESCRIPTION
**Pull request by duncanriach**
# What does this PR do ?
Adds sequence packing support for the hybrid model and also enhances `--sft`, `SFTDataset`, and `pretrain_mamba.py` to feed packed sequence data into the model.

Causes a regression in the functionality of `--sft` on the GPT model; this functionality will no longer be supported. To maintain this functionality, packed-sequence related changes would need to be made to `pretrain_gpt.py` following `pretrain_mamba.py` and, ideally, GPT model SFT packed-sequence convergence testing would need to be run. I believe this functionality will be getting added by PR [2282](https://github.com/NVIDIA/Megatron-LM/pull/2282) from @parthmannan, which will be merged after this current PR.

This PR also does not contain functional tests for packed sequence support in the hybrid model. I believe that there are currently also no functional tests for packed sequence support in the GPT model.

See internal MR 4403

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Phlip79** [comment]
Please fix linting errors.

File: megatron/core/datasets/gpt_dataset.py [hunk 1]
```diff
@@ -55,6 +55,9 @@ class GPTDatasetConfig(BlendedMegatronDatasetConfig):
        Check --per-dataset-sequences-path
     """
 
+    context_parallel_size: Optional[int] = None
+    """The size of the context parallel group. Needed for padding in packed sequences."""
+
```
Comment (asolergi-nv) [inline megatron/core/datasets/gpt_dataset.py:60]: Should we create a SFTDatasetConfig class to control SFTDataset specific args?

File: megatron/training/datasets/sft_dataset.py [hunk 2]
```diff
@@ -25,6 +27,9 @@ class SFTLowLevelDataset:
                 {"role": "user", "content": "something1"},
                 {"role": "assistant", "content": "something2"},
             ]
+            A jsonl line can contain multiple conversations packed together into on list. Each
```
Comment (asolergi-nv) [inline megatron/training/datasets/sft_dataset.py:30]: Is this particular to our dataset format? If we use a dataset where each line contains 1 sample, with this implementation samples will be very short, right?

File: megatron/training/datasets/sft_dataset.py [hunk 3]
```diff
@@ -68,79 +73,131 @@ def build_low_level_dataset(dataset_path: str, config: GPTDatasetConfig) -> LowL
...
+            tokens, targets = tokenizer.tokenize_conversation(
+                conversation, return_target=True, add_generation_prompt=False
+            )
 
-        # Position ids.
-        position_ids = torch.arange(max_seq_len, dtype=torch.long)
+            tokens_list = tokens.tolist()
+            targets_list = targets.tolist()
+
+            # Add EOD, unless it's already present
```
Comment (asolergi-nv) [inline megatron/training/datasets/sft_dataset.py:120]: Why are we forcing the EOD token? Shouldn't this be handled in the tokenizer or chat template?

File: megatron/training/datasets/sft_dataset.py [hunk 4]
```diff
@@ -68,79 +73,131 @@ def build_low_level_dataset(dataset_path: str, config: GPTDatasetConfig) -> LowL
...
+                    extend_with_padding(pack_tokens, pack_targets, pack_positions, pad_len)
+
+            # TODO(duncan): Consider also padding to multiple of number of tokens here. This might
+            # be needed for efficiency (and potentially set via command-line argument).
+
+            cu_seqlens.append(len(pack_tokens))
+
+            # Handle any necessary truncation
+            if len(pack_tokens) >= pack_length + 1:  # +1 here to account for later alignment
+                truncate_left_not_right = True  # TODO(duncan): plumb this switch in
```
Comment (asolergi-nv) [inline megatron/training/datasets/sft_dataset.py:145]: Resolve

File: megatron/training/datasets/sft_dataset.py [hunk 5]
```diff
@@ -68,79 +73,131 @@ def build_low_level_dataset(dataset_path: str, config: GPTDatasetConfig) -> LowL
...
+                    pack_tokens = pack_tokens[-max_body:]
+                    pack_targets = pack_targets[-max_body:]
+                    pack_tokens.append(pad)
+                    pack_targets.append(pad)
+                else:  # Truncate right (need to add eod)
+                    max_body = pack_length - 1
+                    pack_tokens = pack_tokens[:max_body]
+                    pack_targets = pack_targets[:max_body]
+                    pack_tokens.extend([eod, pad])
+                    pack_targets.extend([eod, pad])
```
Comment (asolergi-nv) [inline megatron/training/datasets/sft_dataset.py:157]: Move all padding to later in L163

File: megatron/training/datasets/sft_dataset.py [hunk 6]
```diff
@@ -68,79 +73,131 @@ def build_low_level_dataset(dataset_path: str, config: GPTDatasetConfig) -> LowL
...
-            attention_mask = attention_mask < 0.5
-        else:
-            attention_mask = None
-
-        return loss_mask, position_ids, attention_mask
+        loss_mask = torch.ones(pack_length, dtype=torch.float32)
+        loss_mask[labels == pad] = 0.0  # Mask paddings
+        loss_mask[labels == IGNORE_INDEX] = 0.0  # mask prompts
+
+        # TODO(duncan): Optionally create an attention mask
```
Comment (asolergi-nv) [inline megatron/training/datasets/sft_dataset.py:184]: Resolve

File: pretrain_gpt.py [hunk 7]
```diff
@@ -48,6 +48,10 @@ def get_batch(data_iterator, vp_stage=None):
     # get batches based on the TP rank you are on
     batch = get_batch_on_this_tp_rank(data_iterator)
 
+    assert "cu_seqlens" not in batch and "max_seqlen" not in batch, (
```
Comment (asolergi-nv) [inline pretrain_gpt.py]: Let's check this in `training/arguments.py`

File: megatron/core/datasets/gpt_dataset.py (hunk 1 above)
Comment (duncanriach) [reply megatron/core/datasets/gpt_dataset.py:60]: That might make sense, but perhaps not in this PR

File: megatron/training/datasets/sft_dataset.py (hunk 2 above)
Comment (duncanriach) [reply megatron/training/datasets/sft_dataset.py:30]: Yes, this is a format that we call "fake packed." Yes, if there is only one sample in each line then it will train very inefficiently.

File: megatron/training/datasets/sft_dataset.py (hunk 3 above)
Comment (duncanriach) [reply megatron/training/datasets/sft_dataset.py:120]: I don't really know. I assumed that there would need to always be an EOD, and this seemed to be acceptable to the post-training folks. I know the we've been using identity chat templates, which I believe would not add EOD. I think that perhaps Soumye Singhal could comment.

File: megatron/training/datasets/sft_dataset.py (hunk 4 above)
Comment (duncanriach) [reply megatron/training/datasets/sft_dataset.py:145]: We didn't use this feature for post training of the last several models. I put the code here in case we want to add it, to show how it would be done. I'm not sure we want to plumb this in now, add another command-line flag, and ideally test it. Perhaps I should just strip out the unusable code?

File: megatron/training/datasets/sft_dataset.py (hunk 5 above)
Comment (duncanriach) [reply megatron/training/datasets/sft_dataset.py:157]: I find this code really hard to reason about. I spent a lot of time crafting it, and we've post-trained a lot of models with it. I'm not comfortable removing that pad here.

File: megatron/training/datasets/sft_dataset.py (hunk 6 above)
Comment (duncanriach) [reply megatron/training/datasets/sft_dataset.py:184]: Please can we add this in a later PR?

File: pretrain_gpt.py (hunk 7 above)
Comment (duncanriach) [reply pretrain_gpt.py]: This is the best place to catch this. This is the location where a batch can be returned with these variables that cannot be handled by the later code. This is catching the issue at the place where it would need to be addressed, in `pretrain_gpt.py`. This will also be addressed almost immediately when [2282](https://github.com/NVIDIA/Megatron-LM/pull/2282) is merged after this PR. It will allow this assert to be removed.

**Comment by ericharper** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2913/commits/94aaa5174b7220505699279116128ed077879f7f

**Review by ericharper — APPROVED** [review]
